internal/pr/promote/github_pull_request_provider.go

```diff
@@ -0,0 +1,131 @@
+package promote
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"os/exec"
+	"regexp"
+)
+
+type GitHubPullRequestProvider struct {
+}
+
+var labelRegex = regexp.MustCompile(`^promote:\s*(\w+)$`)
+
+type pullRequest struct {
+	Labels []struct {
+		Name string `json:"name"`
+	}
+}
+
+func get(branch string) (*pullRequest, error) {
+	// List pull requests for branch
+	cmd := exec.Command("gh", "pr", "list", "--head", branch, "--json", "labels")
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		return nil, fmt.Errorf("listing pull requests for branch %s: %s", branch, output)
+	}
+
+	// Unmarshal JSON
+	var prs []pullRequest
+	if err := json.Unmarshal(output, &prs); err != nil {
+		return nil, fmt.Errorf("unmarshaling pull request list: %w", err)
+	}
+
+	// We can safely assume that there is either none or only one PR for a given branch
+	if len(prs) == 0 {
+		return nil, nil
+	}
+	return &prs[0], nil
+}
+
+func (g *GitHubPullRequestProvider) Exists(branch string) (bool, error) {
+	pr, err := get(branch)
+	if err != nil {
+		return false, fmt.Errorf("getting pull request for branch %s: %w", branch, err)
+	}
+	return pr != nil, nil
+}
+
+func (g *GitHubPullRequestProvider) CreateInteractively(branch string) error {
+	cmd := exec.Command("gh", "pr", "create", "--head", branch)
+	cmd.Stdin = os.Stdin
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	if err := cmd.Run(); err != nil {
+		return fmt.Errorf("creating pull request: %w", err)
+	}
+	return nil
+}
+
+func getPromotionLabels(branch string) ([]string, error) {
+	pr, err := get(branch)
+	if err != nil {
+		return nil, err
+	}
+	if pr == nil {
+		return nil, nil
+	}
+	var labels []string
+	for _, label := range pr.Labels {
+		if labelRegex.MatchString(label.Name) {
+			labels = append(labels, label.Name)
+		}
+	}
+	return labels, nil
+}
+
+func (g *GitHubPullRequestProvider) GetPromotionEnvironment(branch string) (string, error) {
+	labels, err := getPromotionLabels(branch)
+	if err != nil {
+		return "", fmt.Errorf("getting promotion labels for branch %s: %w", branch, err)
+	}
+	if len(labels) == 0 {
+		return "", nil
+	}
+	return labelRegex.FindStringSubmatch(labels[0])[1], nil
+}
+
+func (g *GitHubPullRequestProvider) SetPromotionEnvironment(branch string, env string) error {
+	// Get current promotion labels
+	// Typically, there is only one or none, but we cannot guarantee there are not many
+	labels, err := getPromotionLabels(branch)
+	if err != nil {
+		return fmt.Errorf("getting promotion labels for branch %s: %w", branch, err)
+	}
+
+	// Remove existing labels, if any
+	for _, label := range labels {
+		if err := removeLabel(branch, label); err != nil {
+			return fmt.Errorf("removing label %s from branch %s: %w", label, branch, err)
+		}
+	}
```

Comment:
Should we also delete the promotion label on whichever PR may have the same promotion environment?

It could be confusing if two PRs have the same target. Which one would work?

Reply:
Indeed, they would be fighting against each others at every one of their builds.  Per your above comment, let's try to implement that!